rtgui/edit.cc

```diff
@@ -747,7 +747,294 @@ void Rectangle::drawToMOChannel(Cairo::RefPtr<Cairo::Context> &cr, unsigned shor
+        break;
+
+    case (DP_TOPRIGHT):
+        topLeft.x = pos.x - W;
+        topLeft.y = pos.y;
+        break;
+
+    case (DP_CENTERRIGHT):
+        topLeft.x = pos.x - W;
+        topLeft.y = pos.y - H / 2;
+        break;
+
+    case (DP_BOTTOMRIGHT):
+        topLeft.x = pos.x - W;
+        topLeft.y = pos.y - H;
+        break;
+
+    case (DP_BOTTOMCENTER):
+        topLeft.x = pos.x - W / 2;
+        topLeft.y = pos.y - H;
+        break;
+
+    case (DP_BOTTOMLEFT):
+        topLeft.x = pos.x;
+        topLeft.y = pos.y - H;
+        break;
+
+    case (DP_CENTERLEFT):
+        topLeft.x = pos.x;
+        topLeft.y = pos.y - H / 2;
+        break;
+    }
+
+    bottomRight.x = topLeft.x + W - 1;
+    bottomRight.y = topLeft.y + H - 1;
+}
+
+OPIcon::OPIcon(Cairo::RefPtr<Cairo::ImageSurface> *normal,
+               Cairo::RefPtr<Cairo::ImageSurface> *active,
+               Cairo::RefPtr<Cairo::ImageSurface> *prelight,
+               Cairo::RefPtr<Cairo::ImageSurface> *dragged,
+               Cairo::RefPtr<Cairo::ImageSurface> *insensitive,
+               enum DrivenPoint drivenPoint) :
+    drivenPoint(drivenPoint)
+{
+    if (normal) {
+        normalImg = *normal;
+    }
+
+    if (prelight) {
+        prelightImg = *prelight;
+    }
+
+    if (active) {
+        activeImg = *active;
+    }
+
+    if (dragged) {
+        draggedImg = *active;
+    }
+
+    if (insensitive) {
+        insensitiveImg = *insensitive;
+    }
+}
+
+OPIcon::OPIcon(Glib::ustring normalImage, Glib::ustring activeImage, Glib::ustring prelightImage,
+               Glib::ustring  draggedImage, Glib::ustring insensitiveImage, enum DrivenPoint drivenPoint) : drivenPoint(drivenPoint)
+{
+    if (!normalImage.empty()) {
+        normalImg = Cairo::ImageSurface::create_from_png( RTImage::findIconAbsolutePath(normalImage) );
+    }
+
+    if (!prelightImage.empty()) {
+        prelightImg = Cairo::ImageSurface::create_from_png( RTImage::findIconAbsolutePath(prelightImage) );
+    }
+
+    if (!activeImage.empty()) {
+        activeImg = Cairo::ImageSurface::create_from_png( RTImage::findIconAbsolutePath(activeImage) );
+    }
+
+    if (!draggedImage.empty()) {
+        draggedImg = Cairo::ImageSurface::create_from_png( RTImage::findIconAbsolutePath(draggedImage) );
+    }
+
+    if (!insensitiveImage.empty()) {
+        insensitiveImg = Cairo::ImageSurface::create_from_png( RTImage::findIconAbsolutePath(insensitiveImage) );
+    }
+}
+
+Cairo::RefPtr<Cairo::ImageSurface> OPIcon::getNormalImg()
+{
+    return normalImg;
+}
+Cairo::RefPtr<Cairo::ImageSurface> OPIcon::getPrelightImg()
+{
+    return prelightImg;
+}
+Cairo::RefPtr<Cairo::ImageSurface> OPIcon::getActiveImg()
+{
+    return activeImg;
+}
+Cairo::RefPtr<Cairo::ImageSurface> OPIcon::getDraggedImg()
+{
+    return draggedImg;
+}
+Cairo::RefPtr<Cairo::ImageSurface> OPIcon::getInsensitiveImg()
+{
+    return insensitiveImg;
+}
+
+void OPIcon::drawImage(Cairo::RefPtr<Cairo::ImageSurface> img,
```

Comment:
I don't we need to modify or copy the `RefPtr` here, we just want to modify the image itself. So the appropriate type seems to be `const Cairo::RefPtr<Cairo::ImageSurface>&` again? (Copying valid reference-counted pointers usually involves atomic operations on its control block which can become a performance burden.)